Hive Plugins: Add Hive Batch Source and Sink

rohitsinha54 opened the pull request:
- Adding Hive Batch Source
- Tested on HDP 2.2 cluster running Hadoop 2.4 and Hive 0.13 and CDH 5.4.5 running Hadoop 2.6 and Hive 1.1 (with respective dependencies in pom).

Commit:
Hive Plugins: Add Hive Batch Source

anwar6953 commented on hive-plugins/pom.xml:
```diff
@@ -0,0 +1,147 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
```
just make it like this

```
<!--
  Copyright © 2015 Cask Data, Inc.

  Licensed under the Apache License, Version 2.0 (the "License"); you may not
  use this file except in compliance with the License. You may obtain a copy of
  the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
  License for the specific language governing permissions and limitations under
  the License.
  -->
```

anwar6953 commented on hive-plugins/pom.xml:
```diff
@@ -0,0 +1,147 @@
+    </plugins>
+  </build>
+
+</project>
```
needs newline

rohitsinha54 commented:
Note: unit test needs to be added. I will be adding that after I complete the sink.

nitinmotgi commented:
Please add README.rst format same as https://github.com/caskdata/hydrator-plugins/blob/develop/transform-plugins/README.rst before merging.

Commit:
Hive Plugins: Add Hive Batch Sink

Commit:
Add Hive UI rendering JSON file

Commit:
Minor description fix in HiveSinkConfig

Commit:
Fix for fields in HiveBatchSink and minor fixes

Commit:
add an example in metastore URI for HiveConfig

Commit:
Add all modules in default

rohitsinha54 commented:
@anew @gokulavasan : Please review Hive Plugins
- Added Hive Batch Sink
- Refactored code around Hive Plugins for source and sink
- Added Hive json file for UI

Commit:
Add check for schema field to exist in table

albertshau commented on hive-plugins/pom.xml:
```diff
@@ -0,0 +1,146 @@
+              <classifier>batch</classifier>
+              <instructions>
+                <!-- build bundle jar for batch -->
+                <_exportcontents>co.cask.plugin.etl.batch.*;org.apache.commons.lang;org.apache.hive.hcatalog.*;
```
how come this is _exportcontents and not Export-Package?

rohitsinha54 replied:
This is needed to remove duplication in the jar and keep the size smaller. http://felix.apache.org/documentation/subprojects/apache-felix-maven-bundle-plugin-bnd.html#embed-dependency-and-export-package

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -0,0 +1,153 @@
+   * @param schema the {@link Schema}
+   * @return {@link HCatSchema} for the given {@link Schema}
+   */
+  public static HCatSchema toHiveSchema(Schema schema, HCatSchema tableSchema) {
```
im confused by this method. Why do we need to convert the Schema object if we already know the table's schema?

rohitsinha54 replied:
Hive allows projection while reading/writing data. Hive also support some data types which we don't support in our schema. See the getType for detail. If a user want to read a table which have says void type or date or something which we don't support he cannot read it. Doing this way allow user to give a schema where he drops those fields and read the table. See HiveBatchSource for details.

albertshau replied:
ok the idea is to force a schema so that the ui can pass it along....

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -0,0 +1,153 @@
+      case BYTES:
+        return TypeInfoFactory.binaryTypeInfo;
+      default:
+        throw new IllegalArgumentException(String.format(
```
this means nullable types are not supported (nullable is just a union of null and some other type).  I would think we would want to support nullable types.  You can pass in a Schema to this method instead of a Schema.Type, then use the isNullable() method and getNonNullable() at the start of the method.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -0,0 +1,153 @@
+  private static Schema.Type getType(String name, PrimitiveObjectInspector.PrimitiveCategory category) {
+    switch (category) {
+      case BOOLEAN:
+        return Schema.Type.BOOLEAN;
```
don't remember if hive types have a concept of nullable or not.  If so, the schema returned should reflect that. If not, it seems like we should always return a nullable type.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -0,0 +1,153 @@
+   * @param category the field's {@link PrimitiveObjectInspector.PrimitiveCategory}
+   * @return the {@link Schema.Type} for this field
+   */
+  private static Schema.Type getType(String name, PrimitiveObjectInspector.PrimitiveCategory category) {
```
don't remember if hive types have a concept of nullable or not. If so, the schema returned should reflect that. If not, it seems like we should always return a nullable schema.

rohitsinha54 replied:
I looked up and found out that Hive does not have nullable concept for their Schema. So we will return a nullable schema for all.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaStore.java:
```diff
@@ -0,0 +1,53 @@
+
+/**
+ * A class to read/write Hive table schema so that it can be used through various stage of ETL Pipeline.
+ * TODO: This class should be removed once we are able to access information of prepareRun in initialize.
```
the way to do this is to add a setting to the job's Configuration.

albertshau replied:
ok nevermind, you don't have access to Configuration in initialize

albertshau replied:
is there a jira for this?  If not, we should open one.  If so, let's put the jira number in this comment

rohitsinha54 replied:
Yes there is a jira for it and @chtyim is currently working on it. I will add JIRA number todo
https://issues.cask.co/browse/CDAP-4132

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveBatchSink.java:
```diff
@@ -0,0 +1,142 @@
+      hiveSchema = HiveSchemaConverter.toHiveSchema(Schema.parseJson(config.schema), hiveSchema);
+    }
+    HCatOutputFormat.setSchema(configuration, hiveSchema);
+    HiveSchemaStore.storeHiveSchema(context, Objects.firstNonNull(config.dbName, DEFAULT_HIVE_DATABASE),
```
set this in the job Configuration rather than using a dataset.

albertshau replied:
rather than doing this, you should have a no-args constructor for HiveSinkConfig that sets dbName to the default value.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveBatchSink.java:
```diff
@@ -0,0 +1,142 @@
+
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    HCatSchema hCatSchema = HiveSchemaStore.readHiveSchema(context, Objects.firstNonNull(config.dbName,
```
similar comment about no-args constructor for the config.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveBatchSink.java:
```diff
@@ -0,0 +1,142 @@
+    job.setOutputKeyClass(WritableComparable.class);
+    job.setOutputValueClass(DefaultHCatRecord.class);
+    Configuration configuration = job.getConfiguration();
+    configuration.set("hive.metastore.uris", config.metaStoreURI);
```
this should be returned by the output format provider's getOutputFormatConfiguration instead of being set here.  Otherwise you will have issues if 2 hive sinks are used and they connect to different metastores.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveBatchSink.java:
```diff
@@ -0,0 +1,142 @@
+    job.setOutputValueClass(DefaultHCatRecord.class);
+    Configuration configuration = job.getConfiguration();
+    configuration.set("hive.metastore.uris", config.metaStoreURI);
+    HCatOutputFormat.setOutput(configuration, job.getCredentials(), OutputJobInfo.create(Objects.firstNonNull(
```
shouldn't use this method, but should make sure all the settings this sets are reflected in the output format provider's getOutputFormatConfiguration method.  Otherwise this will fail when writing to multiple sinks.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveBatchSink.java:
```diff
@@ -0,0 +1,142 @@
+      // if the user did provide a sink schema to use then use that one
+      hiveSchema = HiveSchemaConverter.toHiveSchema(Schema.parseJson(config.schema), hiveSchema);
+    }
+    HCatOutputFormat.setSchema(configuration, hiveSchema);
```
similar comment here, should be in output format provider

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveBatchSink.java:
```diff
@@ -0,0 +1,142 @@
+      config.dbName, DEFAULT_HIVE_DATABASE), config.tableName, getPartitions()));
+
+    HCatSchema hiveSchema = HCatOutputFormat.getTableSchema(configuration);
+    if (config.schema != null) {
```
why do we need a schema from the user if we have it from hcatalog?

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -0,0 +1,153 @@
+      String name = field.getName();
+      try {
+        // this field of the schema must exist in the table
+        Preconditions.checkNotNull(tableSchema.get(name), "Missing field %s in table schema", name);
```
type should be checked as well

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/commons/HiveSchemaConverter.java:
```diff
@@ -0,0 +1,153 @@
+   * {@link Schema} to exists in the table.
+   *
+   * @param schema the {@link Schema}
+   * @return {@link HCatSchema} for the given {@link Schema}
```
document what gets thrown if the schema is mismatched

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveSinkConfig.java:
```diff
@@ -0,0 +1,44 @@
+ * Configurations for Hive batch sink
+ */
+public class HiveSinkConfig extends HiveConfig {
+  @Name(Hive.FILTER)
```
filter is a misleading name here.  It's not a filter, but the partitions to write to.  'writePartitions' is probably a better name

albertshau replied:
or maybe just 'partitions'

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveSinkConfig.java:
```diff
@@ -0,0 +1,44 @@
+ */
+public class HiveSinkConfig extends HiveConfig {
+  @Name(Hive.FILTER)
+  @Description("Hive expression filter for write provided as a JSON Map of key value pairs that describe all of the " +
```
not an expression filter, but the partition keys and values that will be written to.

albertshau replied:
what happens if I set partitions here, but also have the partition keys in the records I am writing?

rohitsinha54 replied:
Hive will throw an error we dont handle it

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveSinkConfig.java:
```diff
@@ -0,0 +1,44 @@
+    "partition keys and values for that partition. For example if the partition column is 'type' then this property " +
+    "should specified as:" +
+    "\"{\"type\":\"typeOne\"}\"." +
+    "To write multiple partitions simultaneously you can leave this empty, but all of the partitioning columns must " +
```
can I have part of the partition specified here and the other part come from the record?  For example if I'm partitioned by type and date, can type come from this setting and date come from the record?

rohitsinha54 replied:
I don't think its possible as the record is a part of insert into and partition is a part of where clause

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/HiveSinkConfig.java:
```diff
@@ -0,0 +1,44 @@
+
+  @Name(Hive.SCHEMA)
+  @Description("Optional schema to use while writing to Hive table. If no schema is provided then the " +
+    "schema of the source table will be used and it should match the schema of the data being written.")
```
if no schema is provided, the schema of the table will be used (no 'source')

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/RecordToHCatRecordTransformer.java:
```diff
@@ -0,0 +1,59 @@
+   */
+  public HCatRecord toHCatRecord(StructuredRecord record) throws HCatException {
+
+    HCatRecord hCatRecord = new DefaultHCatRecord(record.getSchema().getFields().size());
```
use the size of the this object's schema instead of the record's schema.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/sink/RecordToHCatRecordTransformer.java:
```diff
@@ -0,0 +1,59 @@
+import org.apache.hive.hcatalog.data.schema.HCatSchema;
+
+/**
+ * A transform to convert a {@link StructuredRecord} to Hive's {@link HCatRecord}.
```
document what types of schemas are valid, and the fact that the HCatSchema and Schema objects provided are compatible.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/source/HCatRecordTransformer.java:
```diff
@@ -0,0 +1,72 @@
+import org.apache.hive.hcatalog.data.schema.HCatSchema;
+
+/**
+ * A transform to convert a {@link HCatRecord} from hive to {@link StructuredRecord}.
```
document what types are schemas are valid, and the assumption that the HCatSchema and Schema objects provided are compatible.

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/source/HCatRecordTransformer.java:
```diff
@@ -0,0 +1,72 @@
+
+    for (Schema.Field field : schema.getFields()) {
+      String fieldName = field.getName();
+      switch (field.getSchema().getType()) {
```
should handle nullable fields

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/source/HiveBatchSource.java:
```diff
@@ -0,0 +1,119 @@
+    configuration.set("hive.metastore.uris", config.metaStoreURI);
+    ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+    try {
+      Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
```
add a comment about why this is required

albertshau commented on hive-plugins/src/main/java/co/cask/plugin/etl/batch/source/HiveSourceConfig.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class HiveSourceConfig extends HiveConfig {
+  @Name(Hive.FILTER)
+  @Description("Hive expression filter for scan. This filter must reference only partition columns. " +
```
can't tell what this should be from this description.  Is this like the where clause in a query?

albertshau commented:
took a pass, mostly comments to clarify config settings and to handle schema correctly.  Also, travis checks are failing

Commit:
Add readme and proper locations for resources

sreevatsanraman commented on hive-plugins/README.rst:
```diff
@@ -0,0 +1,127 @@
+You can get started with CDAP-plugins by building directly from the latest source code::
+
+  git clone https://github.com/caskdata/hydrator-plugins.git
+  cd hive-plugins
```
We don't need this cd hive-plugins. We should always build form top level. mvn command won't work otherwise.

Commit:
Addressed comments

rohitsinha54 commented:
Merging this for now. There are some comments which needs to be addressed which I will do in another PR.
